### PR TITLE
Add a visual marker for spaceport mission continuations

### PR DESCRIPTION
**Pull request by bojidar-bg**
**Feature:** This PR implements the feature request detailed and discussed in issue #2712

While the issue has been closed as a duplicate, the feature was not implemented. Feel free to reject this pull request, if the feature is deemed unnecessary.

## Feature Details
This PR adds a mission attribute which would make the map draw a marker for that mission, even if it has not been offered yet. In addition, the Spaceport button is slightly highlighted, so that people don't drop and forget missions by mistake.

Currently, the names of those missions are not displayed in any way. It could potentially be nice to list them somewhere, and perhaps add a textual description to the attribute which would be shown there instead of the actual name.

## UI Screenshots
Without the feature:

| Map | Spaceport button |
| - | - |
| image | image |

With the feature:

| Map | Spaceport button |
| - | - |
| image | image |

## Usage Examples
To use in a mission, just add the ~~`shadow`~~ `reminder` attribute, similar to other existing attributes such as `minor` or `autosave`. It cannot be applied to `boarding` and `assisting` missions, since it requires a single `source` planet to be valid.

I went ahead and manually added the attribute to every mission which had a previous mission mentioning the spaceport in its `"on completed"`.

## Testing Done
I ran through the intro, dropping the mission chain multiple times and observing that the map correctly showed where I can go back to in order to continue it.

## Performance Impact
~~As currently implemented, this might decrease the performance of drawing the map and of moving the mouse over it, due to iterating over all missions to determine which ones have the `shadow` attribute. If considered a problem, it can easily be made faster by storing the available shadow missions and iterating over that collection.~~ Done! Performance should be mostly the same, except when opening the map panel (as we have to go over the available missions to check which ones have reminders available).

### CODE REVIEW

**Comment by Amazinite** [comment]
Related: #1292. The main issue with that PR was that the notification popped up for any spaceport mission, even ones that started a new questline, so this seems like a far better implementation of the idea.

File: source/Mission.cpp [hunk 1]
```diff
@@ -602,6 +606,40 @@ bool Mission::CanOffer(const PlayerInfo &player, const shared_ptr<Ship> &boardin
...
+		auto cit = player.Conditions().find(name + ": offered");
+		if(cit != player.Conditions().end() && cit->second >= repeat)
+			return nullptr;
+	}
+	
+	auto it = actions.find(OFFER);
+	if(it != actions.end() && !it->second.CanBeDone(player, nullptr))
+		return nullptr;
+	
+	it = actions.find(ACCEPT);
```
[the remark this answers is not in the document]
Comment (bojidar-bg) [reply source/Mission.cpp]: This was copied from `::CanOffer`. It probably makes sense, but maybe as a separate PR.

File: source/Mission.cpp [hunk 2]
```diff
@@ -156,6 +156,8 @@ void Mission::Load(const DataNode &node)
 			isVisible = false;
 		else if(child.Token(0) == "priority")
 			hasPriority = true;
+		else if(child.Token(0) == "shadow")
```
[the remark this answers is not in the document]
Comment (bojidar-bg) [reply source/Mission.cpp]: If anyone has an idea for a better attribute, please suggest!
I used "shadow" as this was the first name that popped in my mind.

[the remark this answers is not in the document]
Comment (Amazinite) [reply source/Mission.cpp]: `reminder`? `marked`? `notification`?

[the remark this answers is not in the document]
Comment (bojidar-bg) [reply source/Mission.cpp]: Changed to "reminder".

**Comment by bojidar-bg** [comment]
Synced with master and applied the attribute to the new Nanachi missions.

**Comment by bojidar-bg** [comment]
> I don't think this will work as expected, even for missions with the given attribute, if the missions have any other required parameters or conditions that must be true.

Mission authors are free to add "invisible" missions if adding "reminder" to a visible mission causes things to go awry.
For example, a mission like "Wanderers: Hai Diplomat" can have an invisible reminder, without requiring reminder missions to support location filters, similar to:

```
mission "Wanderers: Hai Diplomat reminder"
	invisible
	landing
	reminder
	source "Ya Hai"
	to offer
		has "First Contact: Wanderer: offered"
		not "Wanderers: Hai Diplomat: offered"
		not "Wanderers: Diplomacy: offered"
	on offer
		fail
```
(I realize it is a bad example due to the existence of "Wanderers: Diplomacy")

> For example, if you need to have X cargo space free, but don't, then the indicator will only show when you have X cargo space free.

This is incorrect. Bunks/crago space is checked in `Mission::HasSpace`, which is not called from `Mission::GetReminderSource`. In this case, when you click the spaceport, you will get the "blocked" message, but the indicator would continue showing.

> Or, worse, if the mission has a random offer chance, each call to [GetReminderSource] will only succeed in accordance with that random proportion: https://youtu.be/HAKNPZIWto4

This is definitelly a problem.  I'm not sure what the best solution to this is, but one option would be to add an argument to `ConditionSet::Test` which forces "random" to either 0 or 100 (or expressions with random to true), while the other would be to add `ConditionSet::HasRandom` and assert that `!toOffer.HasRandom()`.

> The map icon is nice -- "hey, here's an indicator for a mission that you might have forgotten about on some other planet" -- but it can be easily used in an inappropriate manner, and as mentioned above, doesn't actually guarantee that you will get the mission if you go to the spaceport.

I do not understand what "inappropriate manner" you mean, could you elaborate?

Currently, the only reasons to not get the mission when you get to the spaceport would be:
* The use of auto conditions which change when you get to the target planet. (e.g. [#4915](https://github.com/endless-sky/endless-sky/pull/4915))
  * This will still properly show the marker on the spaceport button, but might erroneously show/not show on the map.
* The use of random, which will be fixed.
* Insufficient cargo/bunks space, in which case you would get the "blocked" message.

> Checking every mission in the game also isn't correct, again for that reason, because the mission objects held by GameData are all templates and not actual missions/jobs that the player was actually given / able to accept or decline

One way to do this would be to go over the possible reminder missions and `::Instantiate` each of them (storing them in a list) before trying to show them on the map. I do not see a compelling reason to do so yet, since mission names, descriptions, clearance/blocked messages, bunks/cargo needed, stopovers, waypoints, and destinations (things changed when instantiating) are all not used when displaying a reminder.

**Comment by bojidar-bg** [comment]
Now missions with random `to offer` or `to fail` cannot use `reminder`, thus avoiding the main issue outlined in the review.

**Comment by tehhowch** [comment]
That will fix the visual flicker, yes, but you cannot account for changes in the players conditions (or money, outfits, and anything else required to perform the missions offer, script, decline, and defer actions) between when they see the marker and when they arrive at the respective spaceport.

Showing the marker advertising a spaceport mission and then not offering a mission when the player checks that spaceport is worse than showing no marker at all.

**Comment by bojidar-bg** [comment]
Now caching whether the marker should be shown in the map panel -- I think that performance regressions from this PR should be minimal now.

**Comment by bojidar-bg** [comment]
> That will fix the visual flicker, yes, but you cannot account for changes in the players conditions (or money, outfits, and anything else required to perform the missions offer, script, decline, and defer actions) between when they see the marker and when they arrive at the respective spaceport.

Mission authors are free to apply this attribute at will. If a mission depends on too many conditions (say, date, payment, outfit, and total flagship bunks) they might want to either have no marker (by just not applying the `reminder` attribute) or add a complementary mission which somehow triggers when the player does not have the needed conditions for the main mission.

> Showing the marker advertising a spaceport mission and then not offering a mission when the player checks that spaceport is worse than showing no marker at all.

lndeed. But again, the `reminder` attribute is optional, off by default, and thus mission authors do not have to apply it in their missions if they make such missions. Note that one can already make a mission which ends with `"Please, meet me in the spaceport in an hour"` and then offer no follow up mission. Does it defy player expectations? Yes. Should the engine somehow prevent it? It cannot except with a state-of-art (or better) AI.

Likewise for missions with `reminder` which somehow do not get offered due to some auto condition. We cannot practically simulate all the ways the player can reach the destination planet and decide based on that if a reminder should be shown.

---

I would be happy to do any specific additional changes as needed. Things like renaming attributes and variables, adding some complementary features, fixing specific bugs, etc.

On the other hand, creating a extra-complicated system for handling a few edge case missions using auto conditions, random, and bunks/cargo spaces is a no-go for me. The main usecase for this feature is helping players who accidentally drop the intro mission chain or the Free Worlds plot at some spaceport and then forget which spaceport it was.
So, if you feel that using a system for logic proofs is a must-do for this PR, please close it. I will not be implementing one.

File: source/MapPanel.cpp [hunk 3]
```diff
@@ -1130,6 +1168,7 @@ void MapPanel::DrawTooltips()
 		Point size(hoverText.WrapWidth(), hoverText.Height() - hoverText.ParagraphBreak());
 		size += Point(20., 20.);
 		Point topLeft = (hoverSystem->Position() + center) * Zoom();
+		topLeft += Point(OUTER, OUTER) * Zoom();
```
[the remark this answers is not in the document]
Comment (bojidar-bg) [reply source/MapPanel.cpp]: I suppose. I can revert it, as it is not really related to the PR, but with the heavier tooltip usage related to reminders, it does look better.

**Comment by MasterOfGrey** [comment]
If I have followed the conversation above correctly, it should be fine and intuitive for everything except missions which require X bunks or cargo space. 
And so long as any mission that has a requirement and uses reminder _also_ has a blocked message to explain why, I don't see a problem from a gameplay perspective.

**Comment by petervdmeer** [comment]
I see a big open from the conversation above; tehhowch noticed on Mar 12,2020 that the reminder can show, but that the actual mission might not be offered upon checking the spaceport upon arrival. Promising a player that a mission is available and then silently not delivering is just not acceptable (even if the risk on this happening is low; the game has many players, so this situation will trigger for some players).

@MasterOfGrey The current code doesn't seem to support "blocked messages" for spaceport missions; those missions trigger or they don't.

We might make this somewhat fail-safe by tracking where the pop-up has shown for `"reminder"` missions and to give a message like "based on the map you might have expected a mission to be available here, but no mission could be offered from this spaceport at this time" to the player when he/she lands on the mentioned planet as first stop after seeing the pop-up, checks the spaceport and doesn't find the expected mission. (It is still not very elegant, and I consider it barely acceptable, but it still is better than just silently not delivering.)

I also see a number of other ways to make the code/story-lines themselves more fail-safe;
- Add missions that say "Check the spaceport on XYZ with x cargo-space available" in the mission chains (and those should automatically fail once the relevant mission was offered).
- Add a new feature that allows missions to not only trigger on the spaceport, but trigger on departure. (Instead of "we will meet you at the spaceport", we could have texts like "we will meet you on the landing pad".)
- Rewrite the follow-up missions to automatically trigger (upon departure?)

We also have the regular story reminder option which content creators can use:
```
mission "Diplomat reminder"
	minor
	landing
	destination "planet: Ya Hai"
	to offer
		has "First Contact: Wanderer: offered"
		not "Wanderers: Hai Diplomat: offered"
		not "Wanderers: Diplomacy: offered"
		not "planet: Ya Hai"
	dialog
		"When you land you receive a message; hey captain [last], I am XYZ, a HAI diplomat, that is still waiting for you on Ya Hai."
```
This doesn't need any additional keywords and any content-creator can create custom reminder missions where he/she likes. A similar mechanism already seems to be used [for failed essential freeworlds missions](https://github.com/endless-sky/endless-sky/blob/c8343845a886d0f778729a1e54d723f520c9255a/data/human/free%20worlds%20start.txt#L20).

> So, if you feel that using a system for logic proofs is a must-do for this PR, please close it.

This PR addresses an important annoyance-avoidance feature for many players, so I really like to see it finished. But addressing tehhowch comment (with or without logic proofs) is a must-do; if we cannot find a solution that addresses the comment then closing this PR does seem to be the next logical step.

**Comment by bojidar-bg** [comment]
I'm unfortunately no longer participating in gamedev, so I won't be able to continue my work here, sorry. You are welcome to salvage it or scrap it at your own discretion.

As for the blocked at spaceport issue; it might have been the case that if the current implementation was written so it won't show the reminder at all when the conditions aren't met, thus causing a case of mysteriously disappearing reminder instead of a reminder that doesn't work (note that the reminder shows on the Spaceport button as well, thus providing a small hint even if one doesn't open the map). Showing a blocked message for spaceport missions would be a much better solution though, indeed; but it would be tricky to make it without breaking the "walking around a spaceport and meeting random people" feel.

**Pull request closed without merging**